docs/NEW_NEAR_CLI_INTERFACE.md

```diff
@@ -0,0 +1,114 @@
+# New NEAR CLI interface
+
+`NEAR CLI` is using `extensions` to saticfy all groups of users.
+- `Core NEAR CLI` commands should be usefull for all groups of users.
+- `Core NEAR CLI` is a single binary, there is no extensions that are installed `by default`.
+- Each extension is a separate binary that can be executed from `NEAR CLI`.
+- Extensions are used by a particular group or several groups of users.
+- Extensions are not composable (you should not create extensions for extensions).
+
+`NEAR CLI` is built for:
+- NEAR `dApp developers`, who build smart-contracts, UIs, and tooling on NEAR.
+- `Tech-savvy` people automating their routines.
+- `Validators`
+
+`NEAR CLI UX principles`
+- All altering actions should have a confirmation step with an option to skip confirmation with an explicit command line parameter (e.g. `send` at the end of the command)
+- All direct children commands of a single parent command should be aligned (either represent an action or a resource, but never a mix of those on the same hierarchy level): `contract` -> `state` (resource) and `contract` -> `deploy` (action) are not aligned, so it should be either `contract` -> `get-state` + `contract` -> `deploy` or `contract` -> `state` -> `view` + `contract` -> `code` -> `deploy`
+- Interactive mode should look like: `command - description`. It will help people to learn the commands.
+
+
+## `Core NEAR CLI` commands
+```txt
+import-account
+    - from-near-wallet
+    - from-seed-phrase <seed-phrase>
+    - from-ledger
+    - from-private-key <private-key>
+
+create-account
+    - implicit
+    - subaccount <master-account> <new-subaccount-id>
+
+manage-account <accoundId>
+    - get-state
+    - manage-keys
+        - view
+        - add
+        - delete <public-key>
+    - make-transfer <sender> <reciever> <amount>
+    - manage-contract
+        - deploy-code <wasm-path>
+        - get-code-checksum
+        - get-state
+        - call-view-method
+        - call-change-method
+    - delete
+
+manage-off-chain-keys
+    - ...
+    - generate-key
+    - get-public-key-from-ledger
+
+manage-connections
```

Comment:
Alignment-wise, when I read import-account, create-account, manage-account, manage-keys, and then I see manage-connections I would get confused unless I know that RPCs are involved, and we need to manage those.

I would consider hiding all the CLI configuration commands under a single entrypoint. It seems we can take inspiration from `rustup` tool, which has `rustup self` command to manipulate with its own setup. On another note, we may take inspiration from how `cargo` chooses the compiler version: `cargo build` would use whatever is default, `cargo +beta build` chooses beta version of compiler, so we may have `near +mainnet account manage frol.near keys view`

Reply:
`rustup` has only one nested layer:
```bash
SUBCOMMANDS:
    update          Download and install updates to rustup
    uninstall       Uninstall rustup.
    upgrade-data    Upgrade the internal data format.
    help            Prints this message or the help of the given subcommand(s)
```
Our config and connections are more complex, and `near self connections set mainnet` does not sound right to me. `near connections set mainnet` is more clear. But hiding them under one top level command can be a good idea. Maybe `setup` or `config -> connection` + `config -> behaviour`?

IMO, cargo option is even more confusing for non Rust developers.

Reply:
So far, if we go with resource-oriented top-level commands, `near config connections` looks about right, but I would need to put everything together to see the whole picture before committing to it.

Ok, I don't have a strong opinion about `+network` syntax. Though I still don't see any mention of network selection design in the proposal, so there is nothing to compare with. Should it be something like `--connection mainnet-archival` option at the `near` level?

Reply:
@frol in this design we are not specifying `connection` in a flag or command. We are setting it in config.
```bash
near config connection set mainnet-1
near account manage serhii.near keys list
```

Reply:
@volovyk-s that is going to be super annoying if you want to work with localnet / testnet / mainnet day-to-day. Also, setting a config in one terminal will override the behavior of the command in another terminal. This implicitness is exactly what I hate about near-cli-js. I want to have completely reproducible commands.

Reply:
@frol how about we just have a way to temporarily set the env? We'd still have `near config connection set mainnet-1` as a permanent way to say we want to use mainnet as the default network. But we'd have that other command for the reproducible commands you want. We'd just advise people to use the the temporary config command first, then if they want something more permanent, we'd tell them otherwise

Reply:
@ChaoticTempest do you mean something line `NEAR_ENV` from near-cli-js? I hate it :smile: